alica_engine/include/engine/planselector/PartialAssignmentPool.h

```diff
@@ -1,23 +1,44 @@
-#include <vector>
-
 #pragma once
 
+#include "PartialAssignment.h"
+#include <stdexcept>
+#include <vector>
 namespace alica
 {
-class PartialAssignment;
 class EntryPoint;
 class Task;
 
+class PoolExhaustedException : public std::runtime_error
+{
+public:
+    PoolExhaustedException(const std::string& msg)
+            : std::runtime_error(msg)
+    {
+    }
+};
+
 class PartialAssignmentPool
 {
-  public:
-    PartialAssignmentPool();
-    virtual ~PartialAssignmentPool();
-    int curIndex;
-    const static int maxCount;
-    const EntryPoint* idleEP;
-    const Task* idleTask;
-    std::vector<PartialAssignment*> daPAs;
+public:
+    PartialAssignmentPool(int initialSize);
+    ~PartialAssignmentPool();
+    PartialAssignment* getNext()
+    {
+        PartialAssignment* pa = &_pool[_curIndex];
+
+        if (++_curIndex >= _pool.size()) {
+            throw PoolExhaustedException(std::string("Partial Assignment Pool too small at ") + std::to_string(_pool.size()));
```

Comment:
It would still inherit from std runtime error and need to pass a string. Just for the sake of this single instance I wouldn't do that. Once we have more than one, yes.